src/Components/Components/src/RenderTree/ComponentFrameFlags.cs

```diff
@@ -0,0 +1,17 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Components.RenderTree;
+
+/// <summary>
+/// Types in the Microsoft.AspNetCore.Components.RenderTree namespace are not recommended for use outside
+/// of the Blazor framework. These types will change in future release.
+/// </summary>
+[Flags]
+public enum ComponentFrameFlags : byte
+{
+    /// <summary>
+    /// Indicates that the caller has specified a render mode.
+    /// </summary>
+    HasCallerSpecifiedRenderMode = 1,
+}
```

Comment:
I considered many different approaches for representing the rendermode within the render tree. The eventual approach taken here is:

 * Have the call-site rendermode represented as a new frame type named `ComponentRenderMode`.
   * That gives us flexibility to reference arbitrary objects (which we could not do on an existing `Component` frame), which in turn means that rendermodes can have arbitrary parameters, etc. (which we already want for configuring "prerendering" and "auto mode policy")
 * To minimize the runtime cost, e.g., to avoid the need to scan subtrees for these frames in existing cases where no rendermode was specified, added this new concept of `ComponentFrameFlags` so we can track what extra information exists for a given component frame.
   * This flag is only set when you call `AddComponentRenderMode` at a given call site
   * Only when this flag is set do we scan for the `ComponentRenderMode` frame

So in the default case (i.e., no rendermode) the entire extra cost is simply "check if this flag is set". When it isn't, the rendering and diffing system do nothing extra vs before.

Reply:
One other, more dramatic option would be to eliminate the `@rendermode=...` directive attribute entirely, and say people can *only* declare it on a component-type level. That would simplify/eliminate a lot of implementation.

That still technically covers almost-all circumstances as you can almost-always wrap an extra component around anything in order to specify a type-level attribute. But obviously it's inconvenient. So it comes down to the question of what are the use cases for specifying it at a call site?

For RCL components, we advise them not to have a fixed rendermode if possible. But then when people want to use it interactively, they need one of the following:

 * To be in a non-SSR context already (e.g., a site that's Server globally, or a page that's WebAssembly)
 * Or, to make an artificial wrapper component just to express the rendermode

Maybe we could live with this. The main reason I'm still leaning towards supporting call-site-declared rendermodes is that the teaching/explaining is so much nicer - being able to just add the directive attribute and interactive behaviors light up is super nice. Feedback/opinions welcomed!